### ISSUE TITLE
Mocking cognito issues

### ISSUE THREAD

**Issue by jyenes** [dwyl/aws-sdk-mock#70]
Hi there, 

I'm trying to mock `Cognito` AWS Service in one of the test for my authentication lambda function, here is the piece of code where I'm mocking the service, everything seems fine but in the lambda handler the real service is being called rather than the mocked one, so I'm not sure If I'm doing something wrong or there is a known bug.

## Test
I'm using mocha and chai along with lambda-tester
```javascript
beforeEach(function (done) {
  cognitoMock = AWS.mock('CognitoIdentity', 'getOpenIdTokenForDeveloperIdentity', (params, cb) => {
        cb(null, {
             IdentityId: '12345',
             Token: 'abcd1234'
        });
  });
  done();
});

it('fail: when password is invalid', (done) => {
    LambdaTester(myLambda.handler)
            .event({"body": "{\n \"name\": \"jyenes\",\n \"password\": \"invalid\"\n}"})
            .expectFail( function( err ) {
                expect( err.message ).to
                    .equal('Authentication failed');
            })
            .verify(done);
});
```

## Handler
In the handler I'm calling this method ``getToken`` from the handler after authenticate the user.
```javascript
function getToken(username, cb) {
    const param = {
        IdentityPoolId: config.IDENTITY_POOL_ID,
        Logins: {} 
    };
    param.Logins[config.DEVELOPER_PROVIDER_NAME] = username;
    cognitoidentity.getOpenIdTokenForDeveloperIdentity(param, (err, data) => {
            if (err) cb(err); 
            else cb(null, data.IdentityId, data.Token); 
    });
}
```

Thanks in advance for your help and for your hard work.

**Comment by tfemson** [comment]
@jyenes Please am at the point you were when you raised this issue. Am trying to do a similar thing you did here (Mine too is calling the real service instead of the mocked service)

Did you find any workaround yet?

**Comment by TylerPierro** [comment]
@tfemson Day 489, still no answer from @jyenes.  We wait patiently

**Comment by jyenes** [comment]
> @jyenes Please am at the point you were when you raised this issue. Am trying to do a similar thing you did here (Mine too is calling the real service instead of the mocked service)
> 
> Did you find any workaround yet?

no @tfemson I think I change my approach and did something different, I'm closing this issue

**Comment by mbabauer** [comment]
@jyenes @tfemson : Could either of you share what you did to mock CoginitoIdentity?  I am trying to mock out the authentication but having the same issues mentioned.

**Comment by wmash** [comment]
Unsure if this will help anyone but I was having a similar issue with `CognitoIdentityServiceProvider` whereby the actual service was being called and not the mock. Apparently the problem was that I was requiring the service by object destruction and was, therefore, not initialising properly how the mock has intended. This was my code:

```javascript
// src
const { CognitoIdentityServiceProvider } = require('aws-sdk')

function doSomething() {
    const client = new CognitoIdentityServiceProvider()

    client.adminGetUser(params, function(err, data) {
        // do something
    })
}

// test
const AWSMock = require('aws-sdk-mock')

AWSMock.mock('CognitoIdentityServiceProvider', 'adminGetUser', function(params, cb) {
    cb(null)
}) 
```

When I changed the `src` to:

```javascript
const AWS = require('aws-sdk')

function doSomething() {
    const client = AWS.CognitoIdentityServiceProvider()

    client.adminGetUser(params, function(err, data) {
        // do something
    })
}
```

I also had to explicitly set the SDK instance to mock using so my test file now looks like:

```javascript
const AWS = require('aws-sdk')
const AWSMock = require('aws-sdk-mock')

AWSMock.setSDKInstance(AWS)

AWSMock.mock('CognitoIdentityServiceProvider', 'adminGetUser', function(params, cb) {
    cb(null)
})

/**
 * tests
**/
```

Hope that helps someone :)

**Comment by david-arteaga** [comment]
What I found from testing with this library in general was that generally you will be safe and mocks will work correctly if you explictly call `AWSMock.setSDKInstance(AWS)` always, and do it before you instantiate any other aws service class. So any code of the form `new AWS.SomeService(...)` should always be executed after you set the sdk instance.
I've had lots of issues when starting out but as soon as I started applying these two rules everything went a lot more smoothly.
